tower-balance/src/pool.rs

```diff
@@ -0,0 +1,280 @@
+//! This module defines a load-balanced pool of services that adds new services when load is high.
+//!
+//! The pool uses `poll_ready` as a signal indicating whehter additional services should be spawned
+//! to handle the current level of load. Specifically, every time `poll_ready` on the inner service
+//! returns `Ready`, [`Pool`] consider that a 0, and every time it returns `NotReady`, [`Pool`]
+//! considers it a 1. [`Pool`] then maintains an [exponential moving
+//! average](https://en.wikipedia.org/wiki/Moving_average#Exponential_moving_average) over those
+//! samples, which gives an estimate of how often the underlying service has been ready when it was
+//! needed "recently" (see [`Builder::urgency`]). If the service is loaded (see
+//! [`Builder::loaded_above`]), a new service is created and added to the underlying [`Balance`].
+//! If the service is underutilized (see [`Builder::underutilized_below`]) and there are two or
+//! more services, then the latest added service is removed. In either case, the load estimate is
+//! reset to its initial value (see [`Builder::initial`] to prevent services from being rapidly
+//! added or removed.
+#![deny(missing_docs)]
+
+use super::{Balance, Choose};
+use futures::{Async, Future, Poll};
+use tower_discover::{Change, Discover};
+use tower_service::Service;
+use tower_util::MakeService;
+
+enum Load {
+    /// Load is low -- remove a service instance.
+    Low,
+    /// Load is normal -- keep the service set as it is.
+    Normal,
+    /// Load is high -- add another service instance.
+    High,
+}
+
+/// A wrapper around `MakeService` that discovers a new service when load is high, and removes a
+/// service when load is low. See [`Pool`].
+pub struct PoolDiscoverer<MS, Target, Request>
+where
+    MS: MakeService<Target, Request>,
+{
+    maker: MS,
+    making: Option<MS::Future>,
+    target: Target,
+    load: Load,
+    services: usize,
+}
+
+impl<MS, Target, Request> Discover for PoolDiscoverer<MS, Target, Request>
+where
+    MS: MakeService<Target, Request>,
+    Target: Clone,
+{
+    type Key = usize;
+    type Service = MS::Service;
+    type Error = MS::MakeError;
+
+    fn poll(&mut self) -> Poll<Change<Self::Key, Self::Service>, Self::Error> {
+        if let Load::High = self.load {
+            if self.making.is_none() {
+                try_ready!(self.maker.poll_ready());
+                // TODO: it'd be great if we could avoid the clone here and use, say, &Target
```

Comment:
yeah I think for now this is a pattern we will have to live with 😞

Reply:
In theory, `Target` should be "light"